### DISCUSSION TITLE
How do I contribute to Composer?

### DISCUSSION THREAD

**Discussion by ghost** [composer/composer#11738] [Q&A / Support] [answered]
Let's say I want to contribute to the project. How do I test and debug my changes? I can see three ways:
1. 100% TDD - let's forget it :smile:
2. make changes in composer repository, hardcode composer.(json|lock) in some Composer(Json|Lock)ContentProvider-like class and run Symfony console command as usual
3. use composer repo (cloned or linked) against real app repository instead of PHAR file - is it even possible?

I would like the 3rd option because it would be the easiest and fastest way to move forward with coding in this case.

P.S. I don't have PHP installed on my laptop. I work only in docker environment so it would be very helpful for me if I could just manually debug source code and test changes made in Composer source code with one of my applications. 
Of course I will write unit tests if needed but those I can run on GitHub Actions.

**Comment by Seldaek** [comment, marked as answer]
1. It depends what kind of change you do, some things are very well suited for TDD, others not so.
2. The easiest is to make changes in the git repo, and then run bin/composer with php instead of the phar file, that is how I run composer I never use the phar really.
